internal/service/alarms/internal/db/repo/alarms_repository.go

```diff
@@ -51,3 +60,201 @@ func (ar *AlarmsRepository) GetAlarmEventRecordWithUuid(ctx context.Context, uui
+		sm.Where(
+			psql.Group(psql.Quote("alarm_name"), psql.Quote("resource_type_id")).
+				In(getGetAlertNameAndResourceTypeID(am)...), // Dynamically pass the pairs
+		),
+	)
+
+	sql, params, err := query.Build()
+	if err != nil {
+		return []models.AlarmDefinition{}, fmt.Errorf("failed to build alarm definitions query when processing AM notification: %w", err)
+	}
+
+	// Run query
+	rows, _ := ar.Db.Query(ctx, sql, params...)
+	records, err := pgx.CollectRows(rows, pgx.RowToStructByNameLax[models.AlarmDefinition])
+	if err != nil {
+		return []models.AlarmDefinition{}, fmt.Errorf("failed to collect alarm definitions: %w", err)
+	}
+
+	return records, nil
+}
+
+func getGetAlertNameAndResourceTypeID(am *api.AlertmanagerNotification) []bob.Expression {
+	b := make([]bob.Expression, 0, len(am.Alerts))
+	for _, alert := range am.Alerts {
+		name := alertmanager.GetAlertName(*alert.Labels)
+		resourceTypeId := alertmanager.GetResourceTypeID(*alert.Labels)
+		b = append(b, psql.ArgGroup(name, resourceTypeId))
+	}
+
+	return b
+}
+
+// ArchiveResolvedAlarmEventRecords Delete resolved and move to archive
+func (ar *AlarmsRepository) ArchiveResolvedAlarmEventRecords(ctx context.Context) error {
+	return pgx.BeginFunc(ctx, ar.Db, func(tx pgx.Tx) error { //nolint:wrapcheck
+		query := psql.Delete(
+			dm.From(models.AlarmEventRecordView.NameAs(ctx)),
+			dm.Where(psql.Quote("alarm_status").EQ(psql.Arg(api.Resolved))),
+			dm.Returning(models.AlarmEventRecordView.Columns()),
+		)
+
+		sql, params, err := query.Build()
+		if err != nil {
+			return fmt.Errorf("failed to build delete alarm record query: %w", err)
+		}
+
+		// Run query to delete
+		rows, _ := tx.Query(ctx, sql, params...)
+		records, err := pgx.CollectRows(rows, pgx.RowToStructByNameLax[models.AlarmEventRecord])
+		if err != nil {
+			return fmt.Errorf("failed to delete alarm records: %w", err)
+		}
+
+		// Move deleted data to archived table which is identical to the original. Archived data will later be removed with cronjob.
+		if len(records) > 0 {
+			// Build queries for each record
+			sql, params, err := buildAlarmEventRecordUpsertQuery(records, "alarm_event_record_archive")
+			if err != nil {
+				return fmt.Errorf("failed to build query for event upsert: %w", err)
+			}
+
+			_, err = tx.Exec(ctx, sql, params...)
+			if err != nil {
+				return fmt.Errorf("failed to execute upsert query: %w", err)
+			}
+			slog.Info("Successfully archived alarm event record records", "records", len(records))
+		}
+
+		return nil
+	})
+}
+
+// ResolveNotificationIfNotInCurrent find and only keep the alerts that are available in the current payload
+func (ar *AlarmsRepository) ResolveNotificationIfNotInCurrent(ctx context.Context, am *api.AlertmanagerNotification) error {
+	query := psql.Update(
+		um.Table(models.AlarmEventRecordView.Name(ctx)),
+		um.SetCol("alarm_status").ToArg(api.Resolved),
+		um.Where(psql.Quote("fingerprint").NotIn(getGetAlertFingerPrints(am)...)),
```

Comment:
or if StartsAt is different, it could be a new alert (same fingerprint) that started after again. Assuming, we were off when the previous one went to Resolved and this one is present. Makes sense?

Reply:
Yeah. Update to check for both fingerprint and startAt.